### ISSUE TITLE
Fine tuning on new images - input shape?

### ISSUE THREAD

**Issue by silgonel** [andreped/livermask#11]
Hello again Andre,
I'd like to fine tune your network on MRIs - to see if I can get any good result on liver segmentation on MRI instead of CT.

Can I ask you how the data generator should be organized?
In particular, I am having some issues with input shape of the image and of the mask.
I can see the input image for prediction should be like (1,1,512,512,1), so I guess it's (batchsize, 1 (slice), 512,512, 1 channel)?
It would be super if you could share an example of how the input data is passed to the net.

Thank you very much and have a good weekend,
Silvia

**Comment by andreped** [comment]
The reason for the **odd** input size is actually a mistake of mine. I accidentally added the redundant axis in preprocessing which later on produced a 3DUNet model. If you load the model and look at the layers you will see that there are 3D convolutions there. It is not really a problem, but that is why the input shape is rather odd. I am planning to fix this in the future, but the current model seem to be working well, so I am not in a rush.

Your idea sounds interesting and makes a lot of sense, if you need the model to work on MRIs. However, are you also interested in having a model that works on different sequence types? This might be quite tricky to produce, as it might require a lot of training data, even if you finetune. Thus, I have some pointers for you, before you go down that path:
- If you have corresponding CTs for the MRIs you are working with, you could apply the model on the CTs and register the CT with the MRI to get the liver segmentation on the MRIs. That should work quite well, and should be robust across sequence types. I would recommend using [PyANTs](https://github.com/ANTsX/ANTsPy).
- Other tricks to try would be to apply some type of transform on the CT to make it MRI-like. However, this might be tricky without a deep learning model, e.g. GAN, to do this transform.
- If you have annotations, then no problem, but if you wish to produce a model that is robust across sequence types, you might have to find some augmentations that simulate change in sequence type. That might be tricky, but for liver segmentation that should work well. Does not need to be anything too advanced.

Sadly, the code I used for this project was lost when the desktop I was working on had to be reset, which resulted in loss of data... Hence, I do not have the batch generator I used for this. However, if you are able to share your repository with me, I could help you set it up, such that you are able to train.

The most important thing here, is that you need to preprocess the MRIs, store the preprocessed data on disk, and then use a batch generator and stream the data from disk to the model during training. Assuming you are using TensorFlow, you just need to connect the batch generator with model.fit, as one of the arguments - both for the training and validation set.

However, I have some example code of how this can be done, which I did for another application.

Here is an example of how you connect the batch generator:
https://github.com/andreped/DMDetect/blob/main/python/train_seg.py#L153

And here is an example of a batch generator. It was hand crafted for a segmentation task on mammography images, but the same concept you can use to create your own - streaming MRIs:
https://github.com/andreped/DMDetect/blob/main/python/batch_generator.py#L112

I also have implementations of some segmentation architectures:
https://github.com/andreped/DMDetect/blob/main/python/models.py#L142
https://github.com/andreped/DMDetect/blob/main/python/resunetpp.py#L121

Good luck! :) Let me know if you want me to look at your code.
